File: realtime-whisper-aurix/apps/backend/app/services/embedding_service.py
"""
Cohere Embedding Service
"""

import cohere
from app.core.config import settings
from typing import List


class EmbeddingService:
    """Service for generating embeddings with Cohere"""

    def __init__(self):
        self.client = cohere.Client(settings.COHERE_API_KEY)

    async def generate_embedding(
        self,
        text: str,
        model: str = "embed-english-v3.0",
        input_type: str = "search_document"
    ) -> List[float]:
        """
        Generate embedding for text

        Args:
            text: Text to embed
            model: Cohere embedding model
            input_type: Type of input (search_document or search_query)

        Returns:
            List of float values (embedding vector)
        """
        try:
            response = self.client.embed(
                texts=[text],
                model=model,
                input_type=input_type
            )

            return response.embeddings[0]

        except Exception as e:
            raise Exception(f"Embedding generation failed: {str(e)}")

    async def generate_embeddings_batch(
        self,
        texts: List[str],
        model: str = "embed-english-v3.0",
        input_type: str = "search_document"
    ) -> List[List[float]]:
        """
        Generate embeddings for multiple texts

        Args:
            texts: List of texts to embed
            model: Cohere embedding model
            input_type: Type of input

        Returns:
            List of embedding vectors
        """
        try:
            response = self.client.embed(
                texts=texts,
                model=model,
                input_type=input_type
            )

            return response.embeddings

        except Exception as e:
            raise Exception(f"Batch embedding generation failed: {str(e)}")

File: realtime-whisper-aurix/apps/frontend/electron/embedding-service.ts
import { CohereClient } from 'cohere-ai';

type InputType = 'search_document' | 'search_query';

class EmbeddingService {
  private client: CohereClient;
  private model = 'embed-english-v3.0';

  constructor(apiKey: string) {
    this.client = new CohereClient({ token: apiKey });
  }

  async embed(text: string, inputType: InputType): Promise<number[]> {
    const response = await this.client.embed({
      texts: [text],
      model: this.model,
      inputType,
      embeddingTypes: ['float'],
    });

    const embeddings = response.embeddings as { float?: number[][] };
    if (!embeddings.float || embeddings.float.length === 0) {
      throw new Error('No embeddings returned from Cohere');
    }
    return embeddings.float[0];
  }

  async embedBatch(texts: string[], inputType: InputType): Promise<number[][]> {
    if (texts.length === 0) return [];

    const response = await this.client.embed({
      texts,
      model: this.model,
      inputType,
      embeddingTypes: ['float'],
    });

    const embeddings = response.embeddings as { float?: number[][] };
    if (!embeddings.float) {
      throw new Error('No embeddings returned from Cohere');
    }
    return embeddings.float;
  }
}

let embeddingService: EmbeddingService | null = null;

export function initializeEmbeddingService(apiKey: string): EmbeddingService {
  embeddingService = new EmbeddingService(apiKey);
  console.log('Embedding service initialized');
  return embeddingService;
}

export function getEmbeddingService(): EmbeddingService | null {
  return embeddingService;
}

export async function embedText(text: string, inputType: InputType): Promise<number[]> {
  if (!embeddingService) throw new Error('Embedding service not initialized');
  return embeddingService.embed(text, inputType);
}

export async function embedTexts(texts: string[], inputType: InputType): Promise<number[][]> {
  if (!embeddingService) throw new Error('Embedding service not initialized');
  return embeddingService.embedBatch(texts, inputType);
}
